### ISSUE TITLE
Question about Azure Monitor returning 0 for recent scrapes

### ISSUE THREAD

**Issue by adam-resdiary** [tomkerkhove/promitor#711]
I wanted to follow up on a point that was raised in https://github.com/tomkerkhove/promitor/issues/198 about Azure Monitor returning 0 for the most recent few minutes when using a 1 minute aggregation interval.

At the moment with Azure Metrics Exporter we get a 1 minute aggregation interval for our metrics, and it works because they've taken the approach of arbitrarily asking for a point in time several minutes in the past to work around the limitation in Azure Monitor described in https://github.com/Azure/azure-libraries-for-net/issues/499. Currently, with Promitor we would need to switch to a 5 minute aggregation interval, otherwise we end up with 0 being returned for every point in time, which is obviously not useful to us.

I understand why you don't want to hard-code an arbitrary offset into Promitor to work around this behaviour. But what I wanted to ask is the following:

- Are there any work-arounds for this at the moment, or do we just have to use a 5 minute aggregation at the moment?
- Have you had any thoughts about solving this? For example allowing users to configure an offset or something like that (so we could say always ask for T-2m because we know we won't receive useful information for more recent points in time).

Thanks, and congrats on getting the v1 release out!

**Comment by tomkerkhove** [comment]
Thanks!

That's an interesting approach but the main reason why we are not doing that, is because it could give unexpected results. If this were to be added, I presume that same offset would always be applied, regardless of the aggregation interval then?

What we could do is allow you to define an offset so you could opt-in for it, but certainly not have it turned on by default.

Would that make sense to you?

**Comment by adam-resdiary** [comment]
Yes - I definitely wouldn't want it enabled by default.

But before going down that route, am I right in thinking that we just can't use an aggregation interval of less than 5m at the moment? The reason I'm asking if because I'm not super familiar with the Azure Monitor API, so maybe I've missed something simple we can do to work around it for now.

**Comment by tomkerkhove** [comment]
No that is correct. Azure Monitor has a brief delay which can sometimes take up to ~5 min, right @jsturtevant?

**Comment by adam-resdiary** [comment]
@tomkerkhove I've done some more digging into this, and wanted to share my findings. From what I can see, the result of the 5m interval changes depending on whether Azure Monitor has all the data available or not.

For example, if I have the following results using a 1m interval:

```shell
2019-10-07 14:33:00  CPU percentage  42.11725
2019-10-07 14:34:00  CPU percentage  39.5425
2019-10-07 14:35:00  CPU percentage  36.6995
2019-10-07 14:36:00  CPU percentage  32.38275
2019-10-07 14:37:00  CPU percentage  0
```

I get a 5m average result like the following:

```shell
2019-10-07 14:33:00  CPU percentage  37.6855
```

**NOTE:** the value in the 5m average might be rounded. I've calculated it by hand to create the example, but I've definitely seen this behaviour.

But if I make the same request a few minutes later once it has the results for 14:37, I'll get the following for the 1m average:

```shell
2019-10-07 14:33:00  CPU percentage  42.11725
2019-10-07 14:34:00  CPU percentage  39.5425
2019-10-07 14:35:00  CPU percentage  36.6995
2019-10-07 14:36:00  CPU percentage  32.38275
2019-10-07 14:37:00  CPU percentage  35.684
```

And the following for the 5m average:

```shell
2019-10-07 14:33:00  CPU percentage  37.285199999999996
```

So what I think that means is that if Promitor uses the result for the current time, we'll end up with the wrong value stored in Prometheus because Azure Monitor is calculating it based on partial results.

For reference, I used the following two commands to get the metrics using the Azure CLI:

```shell
az monitor metrics list --resource "/subscriptions/<subscription-id>/resourceGroups/<resource-group>/providers/Microsoft.Sql/servers/<server-name>/databases/<database-name>" --metric "cpu_percent" --start-time "2019-10-06T14:38:00.000Z" --end-time "2019-10-07T14:38:00.000Z" -o table --interval "1m"
az monitor metrics list --resource "/subscriptions/<subscription-id>/resourceGroups/<resource-group>/providers/Microsoft.Sql/servers/<server-name>/databases/<database-name>" --metric "cpu_percent" --start-time "2019-10-06T14:38:00.000Z" --end-time "2019-10-07T14:38:00.000Z" -o table --interval "5m"
```

I think that's pretty similar to what Promitor is doing.

Does that seem correct to you, or am I missing something?

**Comment by tomkerkhove** [comment]
It is very similar indeed, thanks for having a look at this!

We might want to experiment with them and run them in parallel - One with and one without the opt-in switch.

**Comment by tomkerkhove** [comment]
In terms of the opt-in, I think this would make sense:
```yaml
server:
  httpPort: 80 # Optional. Default: 80
azure:
  monitorScrapingWindow:
    enableMovingTimeWindow: true # Optional. Default: false
    timewindow: 5m
```

Those are very bad names but you get the idea I think.

Now given the nature of how scraper works I'm not sure if we can make this an opt-out over time, but we can clearly document this.

**Comment by adam-resdiary** [comment]
Yeah - that makes sense to me. How about `scrapeOffset` as a name? Also, I think `isEnabled` is used elsewhere in that config for bools, so how about:

```yaml
server:
  httpPort: 80 # Optional. Default: 80
azure:
  scrapeOffset:
    isEnabled: true # Optional. Default: false
    offset: 5m
```

Thanks for taking a look at this. Unfortunately because of the way that Azure Monitor seems to return different results depending on the point in time the query is made, it makes this very confusing to understand!

**Comment by tomkerkhove** [comment]
Sounds good to me!

So we should:
- [ ] Update the documentation on who to configure it, why it's there and what the caveat is
- [ ] Update runtime & config to use it
- [ ] Check if we need to change our Helm chart, although I don't think we need it

**Comment by tomkerkhove** [comment]
Is this still an issue? I presume it is?

**Comment by adam-resdiary** [comment]
As far as I know it's still an issue Tom - it's just been low on our priority list so we've just been living with using a 5m aggregation instead of 1m. It's possible that they've made changes to the Azure Monitor API to solve this, but unfortunately I think it was undocumented behaviour, so it's pretty difficult to know without just querying the API and looking at the results you get back. It's very frustrating behaviour with the way the API works!
